src/compton2/Compton2.cc

```diff
@@ -0,0 +1,475 @@
+
+// headers provided in draco:
+#include "compton2/Compton2.hh"
+#include "c4/C4_Functions.hh" //node()
+#include "c4/global.hh"
+#include "ds++/Assert.hh"
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iomanip> // RM
+#include <ios>     // RM
+#include <iostream>
+
+using UINT = size_t;
+using FP = double;
+using vec = std::vector<FP>;
+
+namespace rtt_compton2 {
+
+// xs is assumed to be monotonically increasing and unique
+// returns index in [0,len-2] so that xs[index] and xs[index+1] are valid
+// modifies x so that xs[index] <= x <= xs[index+1]
+UINT find_index(const std::vector<FP> &xs, FP &x) {
+  // Do binary search for index
+  auto loc = std::upper_bound(xs.begin(), xs.end(), x);
+  // loc - begin guaranteed to be in [0,len]
+  UINT index = (loc - xs.cbegin());
+
+  // Move index and x to be interior
+  // len >=2 assumed
+  UINT len = xs.size();
+  FP xmin = xs[0];
+  FP xmax = xs[len - 1U];
+  index = std::min(len - 1U, std::max(UINT(1), index)) - 1U;
+  x = std::max(xmin, std::min(xmax, x));
+  return index;
+}
+
+// xL <= x <= xR assumed
+std::array<FP, 4> hermite(FP x, FP xL, FP xR) {
+  // Spacing of interval
+  FP h = xR - xL;
+  // Left/right basis functions
+  FP bL = (xR - x) / h;
+  FP bR = (x - xL) / h;
+
+  // Hermite functions
+  std::array<FP, 4> H{bL * bL * (3 - 2 * bL), bR * bR * (3 - 2 * bR), -h * bL * bL * (bL - 1),
+                      h * bR * bR * (bR - 1)};
+  return H;
+}
+
+Compton2::Compton2(std::string filename)
+    : num_temperatures_(0U), num_groups_(0U), num_leg_moments_(0U), num_evals_(0U), num_points_(0U),
+      Ts_(0U), Egs_(0U), first_groups_(0U), indexes_(0U), data_(0U), derivs_(0U) {
+  Require(filename.length() > 0U);
+  int rank = rtt_c4::node();
+  constexpr int bcast_rank = 0;
+  int errcode = 0;
+
+  if (rank == bcast_rank) {
+    errcode = read_binary(filename);
+  }
+  broadcast_MPI(errcode);
+
+  Ensure(check_class_invariants());
+}
+
+void Compton2::broadcast_MPI(int errcode) {
+  rtt_c4::global_max(errcode);
+  Insist(errcode == 0, "Non-zero errorcode. Exiting.");
+
+  int rank = rtt_c4::node();
+  constexpr int bcast_rank = 0;
+
+  // Broadcast sizes
+  UINT data_size = data_.size();
+  std::array<UINT, 6> pack = {num_temperatures_, num_groups_, num_leg_moments_,
+                              num_evals_,        num_points_, data_size};
+  rtt_c4::broadcast(&pack[0], pack.size(), bcast_rank);
+  UINT p = 0;
+  num_temperatures_ = pack[p++];
+  num_groups_ = pack[p++];
+  num_leg_moments_ = pack[p++];
+  num_evals_ = pack[p++];
+  num_points_ = pack[p++];
+  data_size = pack[p++];
+
+  // Derived sizes
+  const auto tsz = int(num_temperatures_);
+  const auto egsz = int(num_groups_ + 1U);
+  const auto fgsz = int(num_temperatures_ * num_groups_);
+  const auto isz = int(fgsz + 1U);
+  const auto dsz = int(data_size);
+
+  // Broadcast grids
+  if (rank != bcast_rank)
+    Ts_.resize(tsz);
+  rtt_c4::broadcast(&Ts_[0], tsz, 0);
```

Comment:
Avoid using `&vec[0]` without extra precautions as this is undefined behavior if the vector is empty.

Reply:
See above comment. Not sure the best way to do the broadcast for a vector.